### PR TITLE
Reject surjection proofs with trailing garbage

### PR DESCRIPTION
**Pull request by real-or-random**

### CODE REVIEW

**Comment by apoelstra** [comment]
Can you add a unit test to src/modules/surjection/tests_impl.h with a trailing byte and make sure it's rejecetd?

**Comment by real-or-random** [comment]
Done. I can squash this if desired.

**Comment by apoelstra** [comment]
Looks good! No need to squash. Let me run the tests in valgrind though because I'm a little worried that you don't completely initialize the too-large array and that valgrind will be unhappy about this.

**Comment by instagibbs** [comment]
separate commits for logic/tests are fine

utACK

**Comment by apoelstra** [comment]
ACK

File: src/modules/surjection/tests_impl.h [hunk 1]
```diff
@@ -376,6 +377,12 @@ static void test_gen_verify(size_t n_inputs, size_t n_used) {
     CHECK(secp256k1_surjectionproof_serialize(ctx, serialized_proof, &serialized_len, &proof));
     CHECK(serialized_len == secp256k1_surjectionproof_serialized_size(ctx, &proof));
     CHECK(serialized_len == SECP256K1_SURJECTIONPROOF_SERIALIZATION_BYTES(n_inputs, n_used));
+
+    /* trailing garbage */
+    memcpy(&serialized_proof_trailing, &serialized_proof, serialized_len);
+    serialized_proof_trailing[serialized_len] = seed[0];
```
Comment (real-or-random) [inline src/modules/surjection/tests_impl.h]: I initialize the array here. ;)

Comment (apoelstra) [reply src/modules/surjection/tests_impl.h]: Oh, derp, apparently I can't read :)

**Pull request merged**